### PR TITLE
feat(profiling): create frameIndex and collapse samples

### PR DESCRIPTION
**Pull request by JonasBa**
Adds two utility functions to create a dictionary of frames from profile nodes and the second utility that collapses samples. The first one is very simple, but the general idea behind stack collapsing is that we do not want to process identical sequential stacks by walking each one of them and building the call tree. 

I added code comments, but the general idea is to iterate over all of the samples and try to look ahead for >1 identical samples. As we skip nodes, we update the time delta. If more than 1 sample was skipped, we update the delta and jump our loop to where the skipped position was.

### CODE REVIEW

File: static/app/utils/profiling/guards/profile.tsx [hunk 1]
```diff
@@ -26,7 +26,7 @@ export function isChromeTraceObjectFormat(input: any): input is ChromeTrace.Obje
 // We check for the presence of at least one ProfileChunk event in the trace
 export function isChromeTraceArrayFormat(input: any): input is ChromeTrace.ProfileType {
   return (
-    Array.isArray(input) && input.some(p => p.ph === 'P' && p.name === 'ProfileChunk')
+    Array.isArray(input) && !!input.find(p => p.ph === 'P' && p.name === 'ProfileChunk')
```
Comment (eliashussary) [inline static/app/utils/profiling/guards/profile.tsx:29]: why the change from `.some` to `.find` ?

File: static/app/utils/profiling/profile/chromeTraceProfile.tsx [hunk 2]
```diff
@@ -395,22 +395,139 @@ function collectEventsByProfile(input: ChromeTrace.ArrayFormat): {
...
+    for (let j = 0; j < profileNode.children.length; j++) {
+      const child = frames.get(profileNode.children[j]);
+
+      if (child === undefined) {
+        throw new Error('Missing frame child in profile');
+      }
+
+      child.parent = profileNode;
+    }
+  }
```
Comment (eliashussary) [inline static/app/utils/profiling/profile/chromeTraceProfile.tsx:434]: do we need two loops here? Can't we just move L407 to the 2nd loop?

File: static/app/utils/profiling/profile/chromeTraceProfile.tsx [hunk 3]
```diff
@@ -395,22 +395,139 @@ function collectEventsByProfile(input: ChromeTrace.ArrayFormat): {
...
   _options?: ImportOptions
 ): ProfileGroup {
-  const profiles: Profile[] = [];
+  const {cpuProfiles, threadNames: _threadNames} = collectEventsByProfile(input);
 
-  collectEventsByProfile(input);
+  for (const [_profileId, profile] of cpuProfiles.entries()) {
+    // @ts-ignore
+    // eslint-disable-next-line
+    const index = createFramesIndex(profile);
```
[the remark this answers is not in the document]
Comment (JonasBa) [reply static/app/utils/profiling/profile/chromeTraceProfile.tsx:523]: We'll use it in the following PR's to build the actual call tree

File: static/app/utils/profiling/profile/chromeTraceProfile.tsx [hunk 4]
```diff
@@ -395,22 +395,139 @@ function collectEventsByProfile(input: ChromeTrace.ArrayFormat): {
...
+
+  for (let i = 0; i < profile.nodes.length; i++) {
+    const profileNode = profile.nodes[i];
+
+    if (typeof profileNode.parent === 'number') {
+      const parent = frames.get(profileNode.parent);
+
+      if (parent === undefined) {
+        throw new Error('Missing frame parent in profile');
+      }
```
[the remark this answers is not in the document]
Comment (JonasBa) [reply static/app/utils/profiling/profile/chromeTraceProfile.tsx:418]: These should not happen, if they do, it's fine to have them reported to Sentry. The importing code wraps everything in a try/catch

File: static/app/utils/profiling/guards/profile.tsx (hunk 1 above)
Comment (JonasBa) [reply static/app/utils/profiling/guards/profile.tsx:29]: Oups, not intended

File: static/app/utils/profiling/profile/chromeTraceProfile.tsx [hunk 5]
```diff
@@ -395,22 +395,139 @@ function collectEventsByProfile(input: ChromeTrace.ArrayFormat): {
...
+    const nodeId = profile.samples[i];
+
+    // Initialize the delta to 0, so we can accumulate the deltas of any collapsed samples
+    let delta = 0;
+    // Start at i
+    let j = i;
+    // While we are not at the end and next sample is the same as current
+    while (j < profile.samples.length && profile.samples[j + 1] === nodeId) {
+      // Update the delta and advance j
+      delta = Math.max(delta + profile.timeDeltas[j + 1], delta);
```
Comment (Zylphrex) [inline static/app/utils/profiling/profile/chromeTraceProfile.tsx:488]: Can we leave a comment explaining the negative time deltas and why this isn't just `delta += profile.timeDeltas[j + 1]`.

File: static/app/utils/profiling/profile/chromeTraceProfile.tsx [hunk 6]
```diff
@@ -395,22 +395,139 @@ function collectEventsByProfile(input: ChromeTrace.ArrayFormat): {
...
+      sampleTimes.push(elapsed);
+      samples.push(nodeId);
+      sampleTimes.push(elapsed + delta);
+      elapsed += delta;
+      i = j;
+    } else {
+      // If we have not skipped samples, then we just push the sample and the delta to the list
+      samples.push(nodeId);
+      sampleTimes.push(elapsed);
+      elapsed = Math.max(elapsed + profile.timeDeltas[i + 1], elapsed);
```
Comment (Zylphrex) [inline static/app/utils/profiling/profile/chromeTraceProfile.tsx:507]: Same comment here about the negative time deltas.

**Review by Zylphrex — APPROVED** [review]
